File: 0018 - 4Sum/solution.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        def Nsum(nums, l, r, target, N, now, res):
            if N == 2:
                while l < r:
                    s = nums[l]+nums[r]
                    if s == target:
                        res.append(now+[nums[l], nums[r]])
                        l += 1; r -= 1
                        while l<r and nums[l]==nums[l-1]:
                            l += 1
                        while l<r and nums[r]==nums[r+1]:
                            r -= 1
                    elif s < target:
                        l += 1
                    else:
                        r -= 1
            else:
                for i in range(l, r+1):
                    if i==l or (i>l and nums[i]!=nums[i-1]):
                        Nsum(nums, i+1, r, target-nums[i], N-1, now+[nums[i]], res)
        res = []
        nums.sort()
        Nsum(nums, 0, len(nums)-1, target, 4, [], res)
        return res

File: 0018 - 4Sum/solution.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int n = nums.size();
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        for(int i = 0; i < n-3; i++){
            if(i>0 && nums[i]==nums[i-1]) continue;
            if((long long)nums[i]+nums[i+1]+nums[i+2]+nums[i+3] > target) break;
            if((long long)nums[i]+nums[n-1]+nums[n-2]+nums[n-3] < target) continue;
            for(int j = i+1; j < n-2; j++){
                if(j>i+1 && nums[j]==nums[j-1]) continue;
                if((long long)nums[i]+nums[j]+nums[j+1]+nums[j+2] > target) break;
                if((long long)nums[i]+nums[j]+nums[n-1]+nums[n-2] < target) continue;
                int l = j+1, r = n-1;
                while(l < r){
                    long long s = (long long)nums[i]+nums[j]+nums[l]+nums[r];
                    if(s == target){
                        res.push_back({nums[i], nums[j], nums[l], nums[r]});
                        do{l++;}while(l<r && nums[l]==nums[l-1]);
                        do{r--;}while(l<r && nums[r]==nums[r+1]);
                    }
                    else s<target?(l++):(r--);
                }
            }
        }
        return res;
    }
};